components/prism-php.js

```diff
@@ -139,10 +158,42 @@
 			lookbehind: true
 		},
 		'number': /\b0b[01]+\b|\b0x[\da-f]+\b|(?:\b\d+(?:_\d+)*\.?(?:\d+(?:_\d+)*)*|\B\.\d+)(?:e[+-]?\d+)?/i,
-		'operator': /<?=>|\?\?=?|\.{3}|->|[!=]=?=?|::|\*\*=?|--|\+\+|&&|\|\||[?~]|[/^|%*&<>.+-]=?/,
+		'operator': /<?=>|\?\?=?|\.{3}|\??->|[!=]=?=?|::|\*\*=?|--|\+\+|&&|\|\||<<|>>|[?~]|[/^|%*&<>.+-]=?/,
 		'punctuation': /[{}\[\](),:;]/
 	};
 
+	Prism.languages.insertBefore('php', 'comment', {
+		'attribute': {
+			pattern: /#\[(?:(?!#\[)(?:\\[\s\S]|[^\\\]]))*]/,
```

Comment:
I added some cases to tests and improved attribute pattern.
But i think there are still some edge cases.
(nested attrs definitely will not work with this)

Reply:
> nested attrs definitely will not work with this

Nested attributes aren't supported by PHP yet, I believe, so there's no reason to support them at this point in time.

> But i think there are still some edge cases.

How about this one: `#[Foo("]")]`. This is why I suggested `/#\[(?:["'\[\]]|<php-strings>)*\]/`. 
(Just noticed that I forgot to escape the outer brackets...)

Also, comments.

```php
#[
  // something ]
  Foo,
  /* something else #[] */
  Bar
  # shell comments aren't confusing at all in here
]
```

The full attribute pattern (without nested attributes) is:

```
/#\[(?:[^"'/#\[\]]|\/(?![*/])|("|')(?:\\[\s\S]|(?!\1)[^\\])*\1|(?:\/\/|#(?!\[)).*$|\/\*(?:[^*]|\*(?!\/))*\*\/)*\]/m
```

Easy, right... The whole this is basically a modified `phpPattern`.

Reply:
This is mindblowing :smile:. This pattern is too complex for me. It properly matches comments in attrs. But do not match array `[...]`.
e.g. `#[Attr(3, [2, 3], 'str')]` is not matched.
I can't fix it, it's beyond me.

Reply:
I didn't know you could do `[1,2]` arrays in PHP... This makes things more complicated because we have to count the number of opening and closing brackets which isn't possible with regular expressions. The best we can do is to support a maximum number of nested brackets (e.g. 2 or 3 levels) and call it a day. The resulting pattern will be really long and repetitive, so here is a function that generates it:

```js
function () {
	// <phpExpression>      ::= (?:[^"'/#\[\]]|<division>|<string>|<comment>|#?\[<phpExpression>\])*
	// <string>             ::= "(?:\\[\s\S]|[^\\"])*"|'(?:\\[\s\S]|[^\\'])*'
	// <comment>            ::= <singleline-comment>|<multiline-comment>
	// <singleline-comment> ::= (?:\/\/|#(?!\[)).*$  ; m flag is enabled
	// <multiline-comment>  ::= \/\*(?:[^*]|\*(?!\/))*\*\/
	// <division>           ::= \/(?![*/])  ; this is necessary to match all "/" chars that don't start a comment
	var phpExpression = /(?:[^"'/#\[\]]|\/(?![*/])|"(?:\\[\s\S]|[^\\"])*"|'(?:\\[\s\S]|[^\\'])*'|(?:\/\/|#(?!\[)).*$|\/\*(?:[^*]|\*(?!\/))*\*\/|#?\[<SELF>\])*/.source;
	
	// This will add support for 4 levels of nested [] brackets.
	for (var i = 0; i < 2; i++) {
		phpExpression = phpExpression.replace(/<SELF>/g, function () { return phpExpression; });
	}
	phpExpression = phpExpression.replace(/<SELF>/g, function () { return /[^\s\S]/.source; });
	
	return RegExp('#\\[' + phpExpression + '\\]', 'm');
}
```

Since we now support nested `[]` brackets, I added support for nested attributes as well because it only takes a `#?` before the nested `\[<SELF>\]`-part to do it.

Reply:
Thank you very much for commented function code. I realized that attributes and arrays can appear in different contexts, so I can match attribute with nested array inside without pattern recursion.

Reply:
Your solution is really nice. Let's go with it!